core/src/main/java/org/apache/iceberg/LocationProviders.java

```diff
@@ -68,13 +68,15 @@ public static LocationProvider locationsFor(String location, Map<String, String>
     }
   }
 
+  private static String defaultDataLocation(String tableLocation, Map<String, String> properties) {
+    return properties.getOrDefault(TableProperties.WRITE_NEW_DATA_LOCATION, String.format("%s/data", tableLocation));
```

Comment:
To be honestly,  I did not get what's the real meaning for `TableProperties.WRITE_NEW_DATA_LOCATION`.   If it's the configured default location, then why did it name as `WRITE_NEW_DATA_LOCATION` with its real name as `write.folder-storage.path`.  

Now in this patch,  we are trying to fallback to the `write.folder-storage.path`  when people did not set the `write.object-storage.path`.  It even make this more confuse to understand :-)

Reply:
Yes I totally agree `TableProperties.WRITE_NEW_DATA_LOCATION` is probably not a good name, but it is what it is, we might want to deprecate and rename this if we have a major version update...

The resolution strategy I am trying to create here is that:
- if `write.object-storage.path` is set, use it
- if not found, fallback to `write.folder-storage.path`
- if not found, use `<tableLocation>/data`

Reply:
@openinx any additional thoughts on this?

Reply:
I agree on the fallback strategy, though I'd like to see a warehouse like set up for the whole catalog.

E.g. if I don't set `path`, per table, I'd like for it to work as s3://bucket/warehouse/path/<tableLocation and/or just table name>/<objectstorage_hash>/data

Is that more or less what this config is achieving?

Reply:
Correct. Table location default is defined by the catalog implementation, which is likely `s3://bucket/warehouse/db/tablename`, or a custom override during create table. So that path slash data will always be there when writing data.

Reply:
The main discussion point I think is if we should have `write.folder-storage.path` as a part of the fallback, which I think is worth having, but I do agree it increases the complexity for path resolution which is already quite complicated in Iceberg. If no one else objects the current approach, I will merge this.

Reply:
Or would we wind up getting `s3://bucket/warehouse/<hash>/db/tablename/data`?

My preference would be for the hash to be just after the warehouse for data files if `write.object-storage.path` is not specified, though I can understand that we don't want to complicate the resolution to much. However, keeping the entropy high enough in the path for S3 partitioning seems like a reasonable goal for allowing it to be somewhat convoluted.

I have not made much usage of  `WRITE_NEW_DATA_LOCATION` to have a strong opinion there. I agree the naming is a bit strange (`write.folder-storage.path`), but given that this config already exists (with the given name), I do agree that it would be a bit strange to ignore it entirely if users have set it.

Is there no convention around `write.folder-storage.path`, and `write.folder-storage.*` to only be applied to non-object storage layouts? Given that the property name is `TableProperties.WRITE_NEW_DATA_LOCATION`, I'm guessing that the name is just an arguably unfortunate side effect, but that we should respect the intention of this property.